server/clustering_test.go

```diff
@@ -0,0 +1,577 @@
+				break INNER
+			}
+			if last != expectedLastSeq {
+				time.Sleep(100 * time.Millisecond)
+				break INNER
+			}
+			for i := first; i <= last; i++ {
+				msg, err := store.Lookup(i)
+				if err != nil {
+					t.Fatalf("Error getting message %d: %v", i, err)
+				}
+				if !compareMsg(t, *msg, expectedMsgs[i].data, expectedMsgs[i].sequence) {
+					time.Sleep(100 * time.Millisecond)
+					break INNER
+				}
+			}
+		}
+		return
+	}
+	stackFatalf(t, "Message stores are inconsistent")
+}
+
+func removeServer(servers []*StanServer, s *StanServer) []*StanServer {
+	for i, srv := range servers {
+		if srv == s {
+			servers = append(servers[:i], servers[i+1:]...)
+		}
+	}
+	return servers
+}
+
+func assertMsg(t *testing.T, msg pb.MsgProto, expectedData []byte, expectedSeq uint64) {
+	if !bytes.Equal(msg.Data, expectedData) {
+		stackFatalf(t, "Msg data incorrect, expected: %s, got: %s", expectedData, msg.Data)
+	}
+	if msg.Sequence != expectedSeq {
+		stackFatalf(t, "Msg sequence incorrect, expected: %d, got: %d", expectedSeq, msg.Sequence)
+	}
+}
+
+func compareMsg(t *testing.T, msg pb.MsgProto, expectedData []byte, expectedSeq uint64) bool {
+	if !bytes.Equal(msg.Data, expectedData) {
+		return false
+	}
+	return msg.Sequence == expectedSeq
+}
+
+func TestClusteringConfig(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	opts := GetDefaultOptions()
+	opts.ClusterPeers = []string{"a", "b"}
+	s, err := RunServerWithOpts(opts, nil)
+	if s != nil || err == nil {
+		if s != nil {
+			s.Shutdown()
+		}
+		t.Fatal("Server should have failed to start with cluster peers and no cluster node id")
+	}
+}
+
+// Ensure basic replication works as expected. This test starts three servers
+// in a cluster, publishes messages to the cluster, kills the leader, publishes
+// more messages, kills the new leader, verifies progress cannot be made when
+// there is no leader, then brings the cluster back online and verifies
+// catchup and consistency.
+func TestClusteringBasic(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	// For this test, use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", []string{"b", "c"})
+	s1 := runServerWithOpts(t, s1sOpts, nil)
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", []string{"a", "c"})
+	s2 := runServerWithOpts(t, s2sOpts, nil)
+	defer s2.Shutdown()
+
+	// Configure third server.
+	s3sOpts := getTestDefaultOptsForClustering("c", []string{"a", "b"})
+	s3 := runServerWithOpts(t, s3sOpts, nil)
+	defer s3.Shutdown()
+
+	servers := []*StanServer{s1, s2, s3}
+	for _, s := range servers {
+		checkState(t, s, Clustered)
+	}
+
+	// Create a client connection.
+	sc, err := stan.Connect(clusterName, clientName, stan.PubAckWait(2*time.Second))
```

Comment:
This test often flaps on travis.
Remove the `PubAckWait()` for this connection since this publish is the one creating the channel, so it may take it. Leave it to the default. You are now using a second connection with smaller timeouts for failed cases, so we should be ok.

Reply:
Done